### PR TITLE
Bugfix/aaronjeline/358

### PR DESCRIPTION
**Pull request by aaronjeline**
## Description of changes
Moves our parsers to reject duplicate keys, instead of having last-write-wins semantics.
Adds a new dependency on `serde-with`
## Issue #, if available
#358 and #352 
## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).

I confirm that this PR (choose one, and delete the other options):

- [X] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Does not require updates because my change does not impact the Cedar Dafny model or DRT infrastructure.

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cedar-policy-cli/src/lib.rs [hunk 1]
```diff
@@ -935,25 +928,6 @@ fn read_schema_file(filename: impl AsRef<Path> + std::marker::Copy) -> Result<Sc
         })
 }
 
-fn load_actions_from_schema(entities: Entities, schema: &Option<Schema>) -> Result<Entities> {
```
Comment (aaronjeline) [inline cedar-policy-cli/src/lib.rs:938]: Removing this function has it will be incorporated into the main cedar library, and we want a very specific error semantics for it

File: cedar-policy/tests/example_use_cases_doc.rs [hunk 2]
```diff
@@ -68,9 +68,7 @@ fn scenario_4a() {
 // note: 4b currently omitted because it requires date/timestamp functionality
 
 /// currently failing, as the validator does not support action attributes
-#[should_panic(
-    expected = "error occurred while evaluating policy `policy0`: entity `Action::\\\"view\\\"` does not exist"
-)]
+#[should_panic(expected = "Action::\\\"view\\\"` does not have the attribute `readOnly`")]
```
Comment (aaronjeline) [inline cedar-policy/tests/example_use_cases_doc.rs:71]: Actually having the entity specifies changes the error message.

File: cedar-policy-cli/src/lib.rs (hunk 1 above)
Comment (cdisselkoen) [reply cedar-policy-cli/src/lib.rs:938]: This overlaps with my #360

File: cedar-policy/tests/example_use_cases_doc.rs (hunk 2 above)
Comment (cdisselkoen) [reply cedar-policy/tests/example_use_cases_doc.rs:71]: This overlaps with my #360

Comment (aaronjeline) [reply cedar-policy/tests/example_use_cases_doc.rs:71]: Not surprising. We'll want to make sure yours wins.

Comment (cdisselkoen) [reply cedar-policy/tests/example_use_cases_doc.rs:71]: but I think it's the same change, so we should be able to accept both in either order

File: cedar-policy-cli/src/lib.rs (hunk 1 above)
Comment (cdisselkoen) [reply cedar-policy-cli/src/lib.rs:938]: but I think it's the same change, so we should be able to accept both in either order

File: cedar-policy-core/src/entities/json/entities.rs [hunk 3]
```diff
@@ -417,3 +417,41 @@ impl EntityJSON {
...
+                },
+                "attrs" : {},
+                "parents": []
+            }
+        ]);
+        let eparser: EntityJsonParser<'_, NoEntitiesSchema> =
+            EntityJsonParser::new(None, Extensions::all_available(), TCComputation::ComputeNow);
+        let e = eparser.from_json_value(json).err().unwrap();
+        let bad_euid: EntityUID = r#"User::"alice""#.parse().unwrap();
+        match e {
```
Comment (john-h-kastner-aws) [inline cedar-policy-core/src/entities/json/entities.rs:450]: You can do something nice like
```
assert_matches(e, Err(EntitiesError::Duplicate(euid)) => {
  assert_eq!(bad_euid, euid, r#"Returned euid should be User::"alice""#);
});
```

for this sort of assertion

File: cedar-policy/src/frontend/is_authorized.rs [hunk 4]
```diff
@@ -175,7 +181,9 @@ impl AuthorizationCall {
             None => None,
         };
 
-        let context = Context::from_json_value(self.context, schema.as_ref().map(|s| (s, &action)))
+        let context = serde_json::to_value(self.context)
+            .map_err(|e| [format!("Error encoding the context as JSON: {e}")])?;
```
Comment (john-h-kastner-aws) [inline cedar-policy/src/frontend/is_authorized.rs:185]: This change should be made to the `Context` constructor.

File: cedar-policy/src/frontend/is_authorized.rs [hunk 5]
```diff
@@ -228,62 +277,62 @@ struct RecvdSlice {
     template_instantiations: Option<Vec<TemplateLink>>,
 }
 
-impl RecvdSlice {
```
Comment (john-h-kastner-aws) [inline cedar-policy/src/frontend/is_authorized.rs:231]: Is all this whitespace change to split up the function for the clippy lint?

File: cedar-policy/src/frontend/utils.rs [hunk 6]
```diff
@@ -20,11 +20,13 @@ use std::collections::HashMap;
 
 #[derive(Debug, Serialize, Deserialize)]
 #[serde(untagged)]
+#[serde(expecting = "policy map must have no duplicate IDs,")]
```
Comment (john-h-kastner-aws) [inline cedar-policy/src/frontend/utils.rs:23]: I expect this will give a bad error message if the policy specification happens to be an int or null or something else. Maybe best to spell it out "policies as a concatenated string or multiple policies as a hashmap where the policy Id is the key with no duplicate IDs". Although maybe that's a bit wordy.

File: cedar-policy/src/frontend/is_authorized.rs (hunk 5 above)
Comment (aaronjeline) [reply cedar-policy/src/frontend/is_authorized.rs:231]: To make it more readable, it was a bunch of nested functions.

File: cedar-policy/CHANGELOG.md [hunk 7]
```diff
@@ -22,6 +22,7 @@
 - The `Response::new()` constructor now expects a `Vec<AuthorizationError>` as its third argument.
 - Implements RFC #19, making validation slightly more strict, but more explainable.
 - Improved formatting for error messages.
+- Standardized on duplicates being errors instead of last-write-wins for parsers.
```
Comment (cdisselkoen) [inline cedar-policy/CHANGELOG.md:25]: I think we should be slightly more specific -- maybe list the affected APIs?

File: cedar-policy/src/frontend/is_authorized.rs (hunk 4 above)
Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs:185]: are we serializing here or deserializing here?

**Review by cdisselkoen — APPROVED** [review]

File: cedar-policy/src/frontend/is_authorized.rs (hunk 4 above)
Comment (aaronjeline) [reply cedar-policy/src/frontend/is_authorized.rs:185]: De-serialized to a `map` so that we can use an annotation to fail on duplicates. Before we pass back to Context constructor we have to serialize back to a json-value. If we want to forbid duplicates in the context contructor itself, than this api can't just take a json value.

**Comment by aaronjeline** [comment]
Blocked on: https://github.com/cedar-policy/cedar-java/pull/53

**Review by andrewmwells-amazon — APPROVED** [review]
